dnprogs/fal2/fal.8
.TH FAL 8 "Oct 5 2020" "DECnet utilities"

.SH NAME
fal \- Alternate File Access Listener for DECnet
.SH SYNOPSIS
.B fal
[options]
.br
Options:
.br
[\-dvV]
.SH DESCRIPTION
.PP
.B fal
is a daemon that serves incoming DAP (Data Access protocol) connections from
remote systems. It provides sequential transfer and directory listing from
remote machines using standard DECnet syntax.
It should be started at system boot time (after DECnet has been started) and
must be run as root. Note that this version of
.B fal
does not support decnet.proxy.
.br
The file names output by
.B fal
will always be Unix-style.
.br
The options below affect the behaviour of fal. If you are using
.B dnetd
then these options should be specified in the
.B dnetd.conf(5)
file.
.br
By default all files sent by fal will be sent in STREAMLF format.
.SH OPTIONS
.TP
.I "\-d"
Don't fork and run the background. Use this for debugging.
.TP
.I "\-v"
Verbose. The more of these there are the more verbose fal will be. Don't use
more than one for normal operation because it will seriously impair
performance.
.TP
.I \-h \-?
Displays help for using the command.
.TP
.I \-V
Show the version of fal.


.SH SEE ALSO
.BR dnetd "(8), " dnetd.conf "(5), " dntype "(1), " dndir "(1), " dndel "(1), " dntask "(1), " dnsubmit "(1), " dnprint "(1)"


dnprogs/NEWS
dnprogs (2.61) unstable; urgency=low

  * Include libvaxdata/src/test.c in the distribution tarball
  * Fix password length in DAP connections. It was incorrectly set at 12 
    rather than 40
 
 -- Chrissie Caulfield <chrissie@debian.org>  Sun, 31 Mar 2013 11:18:32 +0000

dnprogs (2.60+x32) unreleased; urgency=low

  * Switch from using deprecated sysctl(2) to writing to /proc/sys.  This
    fixes a build failure on x32.

 -- Daniel Schepler <schepler@debian.org>  Tue, 12 Feb 2013 12:21:31 -0800

dnprogs (2.60) unstable; urgency=low

  * Fixed lots of Debian things that caused lintian problems and 
    upgraded policy version
    Closes: #673408
  * Fix some warnings that might be bugs
  * Incorporate NMU into main tree.

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 26 May 2012 09:54:41 +0100

dnprogs (2.59+nmu2) unstable; urgency=medium

  * NMU
  * dnprogs binaries are Linux-only, so state that. Move dnet-common back
    to binary-all. Closes: #665886

 -- Steve McIntyre <93sam@debian.org>  Mon, 21 May 2012 17:43:14 +0100

dnprogs (2.59+nmu1) unstable; urgency=low

  * NMU
  * Reduce libdnet's recommends on dnet-common to suggests again.
    Closes: #655740. Adapted patch from Jonathan Nieder.

 -- Steve McIntyre <93sam@debian.org>  Fri, 18 May 2012 13:29:53 +0100

dnprogs (2.59) unstable; urgency=low

  * Changed package dnet-common from arch all to any and
    updated dependecies. (Closes: #665886)

 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 27 Apr 2012 09:30:19 +0100

dnprogs (2.58) unstable; urgency=medium

  * Fixed lintian warnings:
    - Fixed changelog entry of last release to have correct line length.
    - Corrected debconf dependency in dnet-common.
    - Corrected versioned dependency in dnprogs on dnet-common.
    - Updated existing overrides to match new syntax.
  * Updated linkage of libdnet_daemon and libdap. (Closes: #646353)
  * Fixed compile warning in dndir when using -Werror=format-security
    (Closes: #646613)
  * Fixed typo in debian/dnet-common.config
  * Do not try to install files into packages on FreeBSD which are
    not build on FreeBSD.
  * Link dapfs with libpthread for new libfuse compatibility
  * Updated nl.po (Closes: #652349)
  * Added pl.po (Closes: #661894)
  * Added it.po (Closes: #663135)
  * Change module-init-tools dependency to kmod (Closes: #662684)

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 10 Mar 2012 10:32:08 +0000

dnprogs (2.57) unstable; urgency=medium

  [ Chrissie Caulfield ]
  * Completed removal of modutils configuration file code
  * Fix presence of CVS directories in tarball
  * updated 'clean' target

  [ Philipp Schafft ]
  * Use /dev/pts PTYs on 64 bit systems (fix makefile bug)
  * Updated translations Debconf templates:
    - German translation (Closes: #615268)
    - French translation (Closes: #615039)
    - Danish translation (Closes: #633984) (added)
    - Slovak translation (Closes: #639629) (added)
  * Look in yet another place for libcrypt (better fix than from NMUs)
  * Added a comment about the detecting of libcrypt. (See: #629664)
  * "ethernet" was capitalized in extended description (Closes: #636798)
  * Fixed some lintian warnings (mainly new rules targets).
  * dnet-common: Do not configure DECnet unless the user requested it.
    (Removed the bad workaround done by NMU) (See: #637179)
  * dnet-common: Added preinst script to clean up the broken config files
    from the last NMU. (Closes: #640861)
  * Upgraded libdnet's suggest to dnet-common to recommend again
    (Closes: #608807)

 -- Chrissie Caulfield <chrissie@debian.org>  Thu,  8 Sep 2011 16:05:47 +0200

dnprogs (2.56.1+nmu1) unstable; urgency=low

  * Non-maintainer upload.
  * libdnet: Only suggest dnet-common package (Closes: #608807)
  * dnet-common: Do not change MAC addresses unless user chooses to
    configure DECnet (Closes: #637179)
  * dnet-common: Restore ability to configure DECnet name and address
    through debconf
  * dnet-common: Remove obsolete modutils configuration file

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 25 Aug 2011 14:26:12 +0100

dnprogs (2.56.1) unstable; urgency=low

  * Non-maintainer upload.
  * Add patch from Ubuntu to fix FTBFS with libcrypt moved to the
    multiarch path.

 -- Aurelien Jarno <aurel32@debian.org>  Tue, 09 Aug 2011 11:58:06 +0200


dnprogs (2.56) unstable; urgency=low

  * Search for libgcrypt in 64bit library paths, too. (thanks to Kamil)
  * Package dnet-common now asks if network interfaces should be configrued
    for DECnet (Closes: #608807)
    Updated translations Debconf templates:
    - Russian translation (Closes: #610642)
    - Portuguese translation (Closes: #611034)
    - Swedish translation (Closes: #611296)
    - Czech translation (Closes: #611430)
    - Japanese translation (Closes: #611448)
    - Spanish translation (Closes: #611547)
  * Build against version 25 of FUSE
    Closes: #598831
  * Fixed some lintian warnings (typos and such).
  * Raised priority of package libdnet from extra to optional
    to comply policy manual section 2.5 "Priorities".

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 22 Feb 2011 10:38:23 +0000

dnprogs (2.55) unstable; urgency=low

  * Build against version 25 of FUSE
    Closes: #598831

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 5 Oct 2010 09:23:12 +0100

dnprogs (2.54) unstable; urgency=low

  * Don't build libvaxdata test program by default.
    Closes: #598752
 
 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 1 Oct 2010 20:44:09 +0100

dnprogs (2.53) unstable; urgency=low

  * Done some more porting to FreeBSD, corrected Build-Depends for hurd-i386
    Closes: #588185
  * Added new tool node(1) - Node name lookup tool

dnprogs/fal2/fal.c
/******************************************************************************
    (C) John Forecast                           john@forecast.name 

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
******************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <unistd.h>
#include <ctype.h>
#include <string.h>
#include <syslog.h>
#include <netdnet/dn.h>
#include <netdnet/dnetdb.h>
#include "dap.h"

#define BUFSIZ          8192
#define MIN(a,b)        (((a) < (b)) ? (a) : (b))

extern int verbosity;

uint8_t sndbuf[BUFSIZ], rcvbuf[BUFSIZ];

/*
 * Parameters extracted from received DAP messages.
 */
uint8_t remOS = 0, remFS = 0, remVersion[5], remSYSCAP[12], remLen = 0,
  remLinuxVMS = 0;
uint16_t bufferSize = BUFSIZ;

uint8_t remDATATYPE[2] = { (1 << DAP_DATATYPE_IMAGE) }, remORG = DAP_ORG_SEQ,
  remRFM = DAP_RFM_FIX, remRAT[3] = { 0 }, remMRS = DAP_MRS_DEFAULT,
  remBKS = DAP_BKS_DEFAULT, remFSZ = DAP_FSZ_DEFAULT, remFOP[6] = { 0 },
  remBSZ = DAP_BSZ_DEFAULT, remDEV[6] = { 0 }, remSDC[6] = { 0 };
uint16_t  remBLS = DAP_BLS_DEFAULT, remDEQ = DAP_DEQ_DEFAULT, remLRL, remFFB;
uint64_t remALQ = DAP_ALQ_DEFAULT, remMRN = DAP_MRN_DEFAULT, remHBK, remEBK,
  remSBN;
  
uint8_t config_seen = 0;

extern void ListDirectory(char *, uint8_t *), OpenFile(char *, uint8_t *),
  CreateFile(char *, uint8_t *), DeleteFile(char *, uint8_t *),
  RenameFile(char *, uint8_t *);
extern int ConnectToFile(void), GetFileData(void), PutFileData(void),
  CloseFile(void);

/*
 * Report an error to the remote system via a Status message
 */
void ReportError(
  uint16_t error
)
{
  void *buf = DAPbuildHeader(DAP_MSG_STATUS);
  
  DAP2bPut(buf, error);
  DAPbPut(buf, 0);
  DAPbPut(buf, 0);
  DAPbPut(buf, 0);

  DAPwrite(buf);
}

/*
 * Process a received Configuration message.
 */
int ProcessConfigurationMessage(void)
{
  void *buf;
  uint16_t bufsz;
  uint8_t *syscap;
  int i;
  
  if (config_seen++ == 0) {
    if (DAP2bGet(&bufsz)) {
      if (bufsz != 0)
        bufferSize = MIN(bufsz, BUFSIZ);

      if (DAPbGet(&remOS) && DAPbGet(&remFS)) {
        for (i = 0; i < sizeof(remVersion); i++)
          if (!DAPbGet(&remVersion[i]))
            return 0;

        if (DAPexValid(&syscap, 12)) {
          if (verbosity > 1)
            DAPlogRcvMessage();
          
          DAPexCopy(syscap, remSYSCAP, sizeof(remSYSCAP));

          if (DAPexGetBit(syscap, DAP_SYSCP_BLOCKING)) {
            remLen = 1;
            if (DAPexGetBit(syscap, DAP_SYSCP_2BYTELEN))
              remLen = 2;
          }
          
          /*
           * We need to determine if the remote system is a real VAX/VMS
           * system or a Linux system pretending to be VAX/VMS. We do this
           * examining the configuration message and looking for specific
           * supported features:
           *
           *    1. Linux sets the buffer size to 0xFFFF, VAX/VMS does not
           *    2. VAX/VMS supports the following options, Linux does not:
           *
           *            A. Relative files
           *            B. Command file submission (by FOP)
           *            C. Segmented messages
           */
          remLinuxVMS = 1;
          if ((bufsz != 0xFFFF) &&
              DAPexGetBit(syscap, DAP_SYSCP_REL) &&
              DAPexGetBit(syscap, DAP_SYSCP_CMDFILE) &&
              DAPexGetBit(syscap, DAP_SYSCP_SEGMENTED))
            remLinuxVMS = 0;
          
          /*
           * Create and send our configuration message.
           *
           * Note that TOPS-10 (and maybe TOPS-20) actually make use of
           * the OS and FS codes to decide how to format directory output.
           * Of course, they have no knowledge of the Linux codes I'm using
           * (I used some reserved for user codes) and default to an
           * unknown O/S and RMS-11 as the file system. If we see the remote
           * system is TOPS-10, we pretend to be Ultrix-32 which the TOPS-10
           * code understands and is able to format directory output
           * correctly.
           */
          buf = DAPbuildHeader(DAP_MSG_CONFIG);
          DAP2bPut(buf, BUFSIZ);
          if (remOS != DAP_OS_TOPS10) {
            DAPbPut(buf, DAP_OS_LINUX);
            DAPbPut(buf, DAP_FS_LINUX);
          } else {
            DAPbPut(buf, DAP_OS_ULTRIX);
            DAPbPut(buf, DAP_FS_ULTRIX);
          }
          DAPbPut(buf, DAP_VER_VERNUM);
          DAPbPut(buf, DAP_VER_ECONUM);
          DAPbPut(buf, DAP_VER_USRNUM);
          DAPbPut(buf, DAP_VER_SOFTVER);
          DAPbPut(buf, DAP_VER_USRSOFT);
          DAPexInit(buf, 12);
          DAPexSetBit(buf, DAP_SYSCP_SEQ);
          DAPexSetBit(buf, DAP_SYSCP_SEQXFER);
          DAPexSetBit(buf, DAP_SYSCP_APPEND);
          DAPexSetBit(buf, DAP_SYSCP_BLOCKING);
          DAPexSetBit(buf, DAP_SYSCP_2BYTELEN);
          DAPexSetBit(buf, DAP_SYSCP_DIRLIST);
          DAPexSetBit(buf, DAP_SYSCP_DATETIME);
          DAPexSetBit(buf, DAP_SYSCP_PROTECT);
//          DAPexSetBit(buf, DAP_SYSCP_SEQREC);
          DAPexSetBit(buf, DAP_SYSCP_RENAME);
          DAPexSetBit(buf, DAP_SYSCP_NAME);
          DAPexDone(buf);

          if (DAPwrite(buf) > 0)
            return 1;
        }
      }
    }
  } else ReportError(DAP_MAC_SYNC | DAP_MIC_SYNC_CONFIG);
  return 0;
}

/*
 * Process a received Access message
 */
int ProcessAccessMessage(void)
{
  uint8_t func, *opt = NULL, *filespec = NULL, *fac = NULL, *shr = NULL;
  uint8_t *display = NULL, *password = NULL;
  char name[256];
  uint8_t len;
  
  if (DAPbGet(&func)) {
    if (DAPexValid(&opt, 5)) {
      if (DAPiGet(&filespec, 255)) {
        if (DAPeom() || DAPexValid(&fac, 3)) {
          if (DAPeom() || DAPexValid(&shr, 3)) {
            if (DAPeom() || DAPexValid(&display, 4)) {
              if (DAPeom() || DAPiGet(&password, 40)) {
                if (verbosity > 1)
                  DAPlogRcvMessage();

                /*
                 * Convert the image format string into a NULL-terminated
                 * string
                 */
                DAPi2string(filespec, name);

                switch (func) {
                  case DAP_AFNC_OPEN:
                    /*
                     * TOPS-10 in some cases will open "*.*" and expect to get
                     * back a list of filenames.
                     */
                    if (remOS == DAP_OS_TOPS10) {
                      if (strchr(name, '*') != NULL) {
                        ListDirectory(name, display);
                        break;
                      }
                    }
                    OpenFile(name, display);
                    break;
                    
                  case DAP_AFNC_CREATE:
                    CreateFile(name, display);
                    break;
                    
                  case DAP_AFNC_RENAME:
                    RenameFile(name, display);
                    break;
                    
                  case DAP_AFNC_ERASE:
                    DeleteFile(name, display);
                    break;
                    
                  case DAP_AFNC_DIRLIST:
                    ListDirectory(name, display);
                    break;
                    
                  case DAP_AFNC_SUBMIT:
                  case DAP_AFNC_EXEC:
                    ReportError(DAP_MAC_UNSUPP | DAP_MIC_ACC_ACCFUNC);
                    return 0;
                }
                return 1;
              }
            }
          }
        }
      }
    }
  }
  return 0;
}

/*
 * Process a received Attributes message
 */
int ProcessAttributesMessage(void)
{
  uint8_t *attmenu = NULL, *datatype = NULL, org, rfm, *rat = NULL,
    *alq = NULL, *mrn = NULL, *runsys = NULL, *fop = NULL, *dev = NULL,
    *sdc = NULL, *hbk = NULL, *ebk = NULL, *sbn = NULL, bks, fsz, bsz;
  uint16_t bls, mrs, deq;
  uint16_t lrl, ffb;
  
  if (DAPexValid(&attmenu, 6)) {
    /*
     * Parse out the fields that are present
     */
    if (DAPexGetBit(attmenu, DAP_ATTR_DATATYPE))
      if (!DAPexValid(&datatype, 2))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_ORG))
      if (!DAPbGet(&org))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_RFM))
      if (!DAPbGet(&rfm))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_RAT))
      if (!DAPexValid(&rat, 3))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_BLS))
      if (!DAP2bGet(&bls))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_MRS))
      if (!DAP2bGet(&mrs))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_ALQ))
      if (!DAPiGet(&alq, 5))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_BKS))
      if (!DAPbGet(&bks))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_FSZ))
      if (!DAPbGet(&fsz))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_MRN))
      if (!DAPiGet(&mrn, 5))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_RUNSYS))
      if (!DAPiGet(&runsys, 40))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_DEQ))
      if (!DAP2bGet(&deq))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_FOP))
      if (!DAPexValid(&fop, 6))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_BSZ))
      if (!DAPbGet(&bsz))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_DEV))
      if (!DAPexValid(&dev, 6))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_SDC))
      if (!DAPexValid(&sdc, 6))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_LRL))
      if (!DAP2bGet(&lrl))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_HBK))
      if (!DAPiGet(&hbk, 5))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_EBK))
      if (!DAPiGet(&ebk, 5))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_FFB))
      if (!DAP2bGet(&ffb))
        return 0;

    if (DAPexGetBit(attmenu, DAP_ATTR_SBN))
      if (!DAPiGet(&sbn, 5))
        return 0;

    if (verbosity > 1)
      DAPlogRcvMessage();

    /*
     * Update the attributes that will be used later.
     */
    if (datatype != NULL) {
      uint64_t tmpDatatype = DAPex2int(datatype);

      /*
       * Validate that only supported datatypes are present and only one
       * data type in selected.
       */
      if (tmpDatatype != 0) {
        uint64_t valid = (1 << DAP_DATATYPE_ASCII) | (1 << DAP_DATATYPE_IMAGE);

        if (((tmpDatatype & ~valid) != 0) || ((tmpDatatype & valid) == valid)) {
          ReportError(DAP_MAC_UNSUPP | DAP_MIC_ATTR_DATATYPE);
          return 1;
        }
      }
      DAPexCopy(datatype, remDATATYPE, sizeof(remDATATYPE));
    }

    if (DAPexGetBit(attmenu, DAP_ATTR_ORG))
      remORG = org;

    if (DAPexGetBit(attmenu, DAP_ATTR_RFM))
      remRFM = rfm;

    if (rat != NULL)
      DAPexCopy(rat, remRAT, sizeof(remRAT));

    if (DAPexGetBit(attmenu, DAP_ATTR_BLS))
      remBLS = bls;

    if (DAPexGetBit(attmenu, DAP_ATTR_MRS))
      remMRS = mrs;

    if (alq != NULL)
      remALQ = DAPi2int(alq);

    if (DAPexGetBit(attmenu, DAP_ATTR_BKS))
      remBKS = bks;

    if (DAPexGetBit(attmenu, DAP_ATTR_FSZ))
      remFSZ = fsz;

    if (mrn != NULL)
      remMRN = DAPi2int(mrn);

    if (DAPexGetBit(attmenu, DAP_ATTR_DEQ))
      remDEQ = deq;

    if (fop != NULL)
      DAPexCopy(fop, remFOP, sizeof(remFOP));

    if (DAPexGetBit(attmenu, DAP_ATTR_BSZ))
      remBSZ = bsz;

    if (dev != NULL)
      DAPexCopy(dev, remDEV, sizeof(remDEV));

    if (sdc != NULL)
      DAPexCopy(sdc, remSDC, sizeof(remSDC));

    if (DAPexGetBit(attmenu, DAP_ATTR_LRL))
      remLRL = lrl;

    if (hbk != NULL)
      remHBK = DAPi2int(hbk);

    if (ebk != NULL)
      remEBK = DAPi2int(ebk);

    if (DAPexGetBit(attmenu, DAP_ATTR_FFB))
      remFFB = ffb;

    if (sbn != NULL)
      remSBN = DAPi2int(sbn);

    return 1;
  }
  return 0;
}

/*
 * Process a received Control message.
 */
int ProcessControlMessage(void)
{
  uint8_t func = DAP_CFNC_GET, *ctlmenu = NULL, rac = 0, *key = NULL,
    krf = DAP_KRF_PRIMARY, *rop = NULL, *hsh = NULL, *display = NULL;

  if (DAPbGet(&func)) {
    if (DAPexValid(&ctlmenu, 4)) {
      /*
       * Parse out the fields that are present
       */
      if (DAPexGetBit(ctlmenu, DAP_CMNU_RAC))
        if (!DAPbGet(&rac))
          return 0;

      if (DAPexGetBit(ctlmenu, DAP_CMNU_KEY))
        if (!DAPiGet(&key, 255))
          return 0;

      if (DAPexGetBit(ctlmenu, DAP_CMNU_KRF))
        if (!DAPbGet(&krf))
          return 0;

      if (DAPexGetBit(ctlmenu, DAP_CMNU_ROP))
        if (!DAPexValid(&rop, 6))
          return 0;

      if (DAPexGetBit(ctlmenu, DAP_CMNU_HSH))
        if (!DAPiGet(&hsh, 5))
          return 0;

      if (DAPexGetBit(ctlmenu, DAP_CMNU_DISPLAY))
        if (!DAPexValid(&display, 4))
          return 0;
    }
  }

  if (verbosity > 1)
    DAPlogRcvMessage();

  switch (func) {
    case DAP_CFNC_GET:
      return GetFileData();
      
    case DAP_CFNC_CONNECT:
      return ConnectToFile();
      
    case DAP_CFNC_PUT:
      return PutFileData();
      
    case DAP_CFNC_UPDATE:
    case DAP_CFNC_DELETE:
    case DAP_CFNC_REWIND:
    case DAP_CFNC_TRUNCATE:
    case DAP_CFNC_MODIFY:
    case DAP_CFNC_RELEASE:
    case DAP_CFNC_FREE:
    case DAP_CFNC_FLUSH:
    case DAP_CFNC_NXTVOL:
    case DAP_CFNC_FIND:
    case DAP_CFNC_EXTEND:
    case DAP_CFNC_DISPLAY:
    case DAP_CFNC_FORWARD:
    case DAP_CFNC_BACKWARD:
      ReportError(DAP_MAC_UNSUPP | DAP_MIC_CTL_CTLFUNC);
  }
  return 0;
}

/*
 * Process a received Access Complete message.
 */
int ProcessAccessCompleteMessage(void)
{
  uint8_t func, *fop = NULL;
  uint16_t check;
  void *buf;
  
  if (DAPbGet(&func)) {
    if (DAPeom() || DAPexValid(&fop, 6)) {
      if (DAPeom() || DAP2bGet(&check)) {

        if (verbosity > 1)
          DAPlogRcvMessage();

        switch (func) {
          case DAP_AFNC_CLOSE:
            return CloseFile();
            
          case DAP_AFNC_ENDSTREAM:
            buf = DAPbuildHeader(DAP_MSG_COMPLETE);
            DAPbPut(buf, DAP_AFNC_RESPONSE);
            DAPwrite(buf);
            return 1;
            
          case DAP_AFNC_RESPONSE:
          case DAP_AFNC_PURGE:
          case DAP_AFNC_SKIP:
            ReportError(DAP_MAC_UNSUPP | DAP_MIC_COMP_CMPFUNC);
        }
      }
    }
  }
  return 0;
}

/*
 * Process a received Protection attributes extension message
 */
int ProcessProtectionAttributesMessage(void)
{
  uint8_t *protmenu = NULL, *owner = NULL, *protsys = NULL, *protown = NULL,
    *protgrp = NULL, *protwld = NULL;

  if (DAPexValid(&protmenu, 6)) {
    /*
     * Parse out the fields that are present
     */
    if (DAPexGetBit(protmenu, DAP_PMNU_OWNER))
      if (!DAPiGet(&owner, 40))
        return 0;

    if (DAPexGetBit(protmenu, DAP_PMNU_PROTSYS))
      if (!DAPexValid(&protsys, 3))
        return 0;

    if (DAPexGetBit(protmenu, DAP_PMNU_PROTOWN))
      if (!DAPexValid(&protown, 3))
        return 0;

    if (DAPexGetBit(protmenu, DAP_PMNU_PROTGRP))
      if (!DAPexValid(&protgrp, 3))
        return 0;

    if (DAPexGetBit(protmenu, DAP_PMNU_PROTWLD))
      if (!DAPexValid(&protwld, 3))
        return 0;

    if (verbosity > 1)
      DAPlogRcvMessage();
    
    /*
     * Update the protection codes that will be used later
     */
    /*** TODO ***/
    return 1;
  }
  return 0;
}

/*
 * Process a received Date and Time attributes extension message
 */
int ProcessDateTimeAttributesMessage(void)
{
  char cdtString[20] = { 0 }, rdtString[20] = { 0 }, edtString[20] = { 0 };
  uint8_t *datmenu = NULL, *cdt = NULL, *rdt = NULL, *edt = NULL;
  uint16_t rvn;
  
  if (DAPexValid(&datmenu, 6)) {
    if (DAPexGetBit(datmenu, DAP_DMNU_CDT))
      if (!DAPfGet(&cdt, 18))
        return 0;

    if (DAPexGetBit(datmenu, DAP_DMNU_RDT))
      if (!DAPfGet(&rdt, 18))
        return 0;

    if (DAPexGetBit(datmenu, DAP_DMNU_EDT))
      if (!DAPfGet(&edt, 18))
        return 0;

    if (DAPexGetBit(datmenu, DAP_DMNU_RVN))
      if (!DAP2bGet(&rvn))
        return 0;

    if (cdt != NULL)
      DAPf2string(cdt, 18, cdtString);
    if (rdt != NULL)
      DAPf2string(rdt, 18, rdtString);
    if (edt != NULL)
      DAPf2string(edt, 18, edtString);

    if (verbosity > 1)
      DAPlogRcvMessage();
    
    return 1;
  }
  return 0;
}

/*
 * Process requests from a socket until we get an error, the socket is closed
 * by the other end or we have completed processing. This routine handles the
 * first-level message processing which can initiate an operation.
 */
void process_request(
  int sock
)
{
  int status;
  
  DAPinit(sock, rcvbuf, BUFSIZ);
  
  /*
   * Wait until the first message is available from the client
   */
  status = DAPread();

  if ((status != -1) && (status != 0)) {
    for (;;) {
      uint8_t type;

      status = DAPnextMessage(&type);
      if ((status == -1) || (status == 0))
        break;

      switch (type) {
        case DAP_MSG_CONFIG:
          if (ProcessConfigurationMessage() <= 0)
            goto fail;
          break;
          
        case DAP_MSG_ACCESS:
          if (ProcessAccessMessage() <= 0)
            goto fail;
          break;
          
        case DAP_MSG_ATTRIBUTES:
          if (ProcessAttributesMessage() <= 0)
            goto fail;
          break;
          
        case DAP_MSG_CONTROL:
          if (ProcessControlMessage() <= 0)
            goto fail;
          break;
          
        case DAP_MSG_COMPLETE:
          if (ProcessAccessCompleteMessage() <= 0)
            goto fail;
          break;
          
        case DAP_MSG_PROTECT:
          if (ProcessProtectionAttributesMessage() <= 0)
            goto fail;
          break;
          
        case DAP_MSG_DATETIME:
          if (ProcessDateTimeAttributesMessage() <= 0)
            goto fail;
          break;
          
        case DAP_MSG_CONTINUE:
        case DAP_MSG_ACKNOWLEDGE:
        case DAP_MSG_DATA:
        case DAP_MSG_STATUS:
        case DAP_MSG_KEYDEF:
        case DAP_MSG_ALLOC:
        case DAP_MSG_SUMMARY:
        case DAP_MSG_NAME:
        case DAP_MSG_ACL:
          if (verbosity > 1)
            DAPlogUnexpectedMessage();
          
          ReportError(DAP_MAC_SYNC | type);
          break;
          
        default:
          ReportError(DAP_MAC_SYNC | DAP_MIC_SYNC_UNKNOWN);
          break;
      }
    }
  }
 fail:
  close(sock);
}
